hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -1709,6 +1700,9 @@ protected HFileBlock readBlockDataInternal(FSDataInputStream is, long offset,
         if (verifyChecksum && !validateChecksum(offset, curBlock, hdrSize)) {
           return null;
         }
+        // remove checksum from buffer now that it's verified
+        int sizeWithoutChecksum = curBlock.getInt(Header.ON_DISK_DATA_SIZE_WITH_HEADER_INDEX);
+        curBlock = curBlock.duplicate().limit(sizeWithoutChecksum);
```

Comment:
Do we need to duplicate here? The old instance could be reference by others?

Reply:
Thanks for reviewing and good catch. I just pushed a fix for that. I also added a few more assertions to my new test